File: openc3/lib/openc3/utilities/metric.rb
# encoding: ascii-8bit

# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.
#
# Note: This file has been completely reimplemented post version 5.3.0

require 'openc3/models/metric_model'

module OpenC3
  class Metric
    # The update interval. How often in seconds metrics are updated by this process
    UPDATE_INTERVAL = 5

    # Mutex protecting class variables
    @@mutex = Mutex.new

    # Array of instances used to keep track of metrics
    @@instances = []

    # Thread used to post metrics across all classes
    @@update_thread = nil

    # Sleeper used to delay update thread
    @@update_sleeper = nil

    # Objects with a generate method to be called on each metric cycle (to generate metrics)
    @@update_generators = []

    attr_reader :microservice
    attr_reader :scope
    attr_reader :data
    attr_reader :mutex

    def initialize(microservice:, scope:)
      @scope = scope
      @microservice = microservice
      @data = {}
      @mutex = Mutex.new

      # Always make sure there is a update thread
      @@mutex.synchronize do
        @@instances << self

        unless @@update_thread
          @@update_thread = OpenC3.safe_thread("Metrics") do
            update_thread_body()
          end
        end
      end
    end

    def set(name:, value:, type: nil, unit: nil, help: nil, labels: nil, time_ms: nil)
      @mutex.synchronize do
        @data[name] ||= {}
        @data[name]['value'] = value
        @data[name]['type'] = type if type
        @data[name]['unit'] = unit if unit
        @data[name]['help'] = help if help
        @data[name]['labels'] = labels if labels
        @data[name]['time_ms'] = time_ms if time_ms
      end
    end

    def set_multiple(data)
      @mutex.synchronize do
        @data.merge!(data)
      end
    end

    def update_thread_body
      @@update_sleeper = Sleeper.new
      while true
        start_time = Time.now

        @@mutex.synchronize do
          @@update_generators.each do |generator|
            generator.generate(@@instances[0])
          end

          @@instances.each do |instance|
            instance.mutex.synchronize do
              json = {}
              json['name'] = instance.microservice
              values = instance.data
              json['values'] = values
              MetricModel.set(json, scope: instance.scope) if values.length > 0
            end
          end
        end

        # Only check whether to update at a set interval
        run_time = Time.now - start_time
        sleep_time = UPDATE_INTERVAL - run_time
        sleep_time = 0 if sleep_time < 0
        break if @@update_sleeper.sleep(sleep_time)
      end
    end

    def shutdown
      @@mutex.synchronize do
        @@instances.delete(self)
        if @@instances.length <= 0
          @@update_sleeper.cancel if @@update_sleeper
          OpenC3.kill_thread(self, @@update_thread) if @@update_thread
          @@update_thread = nil
        end
      end
    end

    def graceful_kill
    end

    def self.add_update_generator(object)
      @@update_generators << object
    end
  end
end

begin
  require 'openc3-enterprise/utilities/metric'
rescue LoadError
  # Open Source Edition - Do nothing here
end

File: openc3/python/openc3/utilities/metric.py
# Copyright 2025 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import time
import threading
from openc3.models.metric_model import MetricModel
from openc3.utilities.sleeper import Sleeper
from openc3.top_level import kill_thread


class Metric:
    # The update interval. How often in seconds metrics are updated by this process
    UPDATE_INTERVAL = 5

    # Mutex protecting class variables
    mutex = threading.Lock()

    # Array of instances used to keep track of metrics
    instances = []

    # Thread used to post metrics across all classes
    update_thread = None

    # Sleeper used to delay update thread
    update_sleeper = None

    # Objects with a generate method to be called on each metric cycle (to generate metrics)
    update_generators = []

    def __init__(self, microservice, scope):
        self.scope = scope
        self.microservice = microservice
        self.data = {}
        self.mutex = threading.Lock()

        # Always make sure there is a update thread
        with Metric.mutex:
            Metric.instances.append(self)
            if Metric.update_thread is None:
                Metric.update_thread = threading.Thread(
                    target=self.update_thread_body,
                    daemon=True,
                )
                Metric.update_thread.start()

    def set(self, name, value, type=None, unit=None, help=None, labels=None, time_ms=None):
        with self.mutex:
            if self.data.get(name) is None:
                self.data[name] = {}
            self.data[name]["value"] = value
            if type:
                self.data[name]["type"] = type
            if unit:
                self.data[name]["unit"] = unit
            if help:
                self.data[name]["help"] = help
            if labels:
                self.data[name]["labels"] = labels
            if time_ms:
                self.data[name]["time_ms"] = time_ms

    def set_multiple(self, data):
        with self.mutex:
            self.data = self.data | data

    def update_thread_body(self):
        Metric.update_sleeper = Sleeper()
        while True:
            start_time = time.time()

            with Metric.mutex:
                for generator in Metric.update_generators:
                    generator.generate(Metric.instances[0])

                for instance in Metric.instances:
                    with instance.mutex:
                        json = {}
                        json["name"] = instance.microservice
                        values = instance.data
                        json["values"] = values
                        if len(values) > 0:
                            MetricModel.set(json, scope=instance.scope)

            # Only check whether to update at a set interval
            run_time = time.time() - start_time
            sleep_time = Metric.UPDATE_INTERVAL - run_time
            if sleep_time < 0:
                sleep_time = 0
            if Metric.update_sleeper.sleep(sleep_time):
                break

    def shutdown(self):
        with Metric.mutex:
            try:
                Metric.instances.remove(self)
            except ValueError:
                pass
            if len(Metric.instances) <= 0:
                if Metric.update_sleeper:
                    Metric.update_sleeper.cancel()
                if Metric.update_thread:
                    kill_thread(self, Metric.update_thread)
                Metric.update_thread = None

    def graceful_kill(self):
        pass

    @classmethod
    def add_update_generator(cls, object):
        Metric.update_generators.append(object)
